File: sajha_menu/sajha_menu/page/bar_order/bar_order.py
import frappe
from datetime import datetime
# from jinja2 import Template


# def clever_function():
#     return "Hello"


# template = Template("{{clever_function}}")
# template.globals['clever_function'] = clever_function


@frappe.whitelist(allow_guest=True)
def getTableName():
    table_name = frappe.db.sql("""SELECT name from `tabTables`""")
    return table_name


@frappe.whitelist()
def getBodyHTML(fromDate, toDate):
    fromDate = datetime.strptime(fromDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    fromDate = datetime.strptime(fromDate, '%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%d-%m-%Y')

    bot_order = frappe.db.get_list("Bar Orders", filters=[
                                   ['created_date', 'between', [fromDate, toDate]]])

    botData = []
    for bot in bot_order:
        active = 0
        done = 0
        orders = frappe.get_doc("Bar Orders", bot.name)
        for order in orders.items:
            if order.order_status == 0:
                active = active + 1
            else:
                done = done + 1

        botData.append(
            {"name": orders.table_name, 'bot_status': orders.bot_status, "status": orders.status, "activeOrder": active, "doneOrder": done, "totalOrder": len(orders.items), "bot_number": orders.bot_id, "bot_name": bot.name})
    return frappe.render_template('sajha_menu/page/templates/bot_orders.html', {'bots': botData})


@frappe.whitelist()
def on_update():
    frappe.publish_realtime("kot_bot_refresh")


@frappe.whitelist()
def applyFilter(company, fromDate, toDate, status):

    if status == "":
        status = None

    fromDate = datetime.strptime(fromDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    fromDate = datetime.strptime(fromDate, '%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%d-%m-%Y')

    if company == "":
        bot_order = frappe.db.get_list("Bar Orders", filters=[
            ['created_date', 'between', [fromDate, toDate]], ['status', '=', status]])
    else:
        bot_order = frappe.db.get_list("Bar Orders", filters=[
            ['created_date', 'between', [fromDate, toDate]], ['company', 'like', company], ['status', '=', status]])

    botData = []
    for bot in bot_order:
        active = 0
        done = 0
        orders = frappe.get_doc("Bar Orders", bot.name)
        for order in orders.items:
            if order.order_status == 0:
                active = active + 1
            else:
                done = done + 1

        botData.append(
            {"name": orders.table_name, 'bot_status': orders.bot_status, "status": orders.status, "activeOrder": active, "doneOrder": done, "totalOrder": len(orders.items), "bot_number": orders.bot_id, "bot_name": bot.name})
    return frappe.render_template('sajha_menu/page/templates/bot_orders.html', {'bots': botData})


@frappe.whitelist()
def getCompanyList():
    names = []
    company = frappe.get_list("Company", fields=['name'])
    print(company)
    for name in company:
        names.append(name['name'])
    return names


@frappe.whitelist()
def refreshKot(company, fromDate, toDate, status):
    fromDate = datetime.strptime(fromDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    fromDate = datetime.strptime(fromDate, '%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%Y-%m-%d').strftime('%d-%m-%Y')
    toDate = datetime.strptime(toDate, '%d-%m-%Y')

    if company == "" or company == None:
        bot_order = frappe.db.get_list("Bar Orders", filters=[
            ['created_date', 'between', [fromDate, toDate]], ['status', '=', status]])
    else:
        bot_order = frappe.db.get_list("Bar Orders", filters=[
            ['created_date', 'between', [fromDate, toDate]], ['company', 'like', company], ['status', '=', status]])

    botData = []
    for bot in bot_order:
        active = 0
        done = 0
        orders = frappe.get_doc("Bar Orders", bot.name)
        for order in orders.items:
            if order.order_status == 0:
                active = active + 1
            else:
                done = done + 1

        botData.append(
            {"name": orders.table_name, 'bot_status': orders.bot_status, "status": orders.status, "activeOrder": active, "doneOrder": done, "totalOrder": len(orders.items), "bot_number": orders.bot_id, "bot_name": bot.name})
    return frappe.render_template('sajha_menu/page/templates/bot_orders.html', {'bots': botData})

File: sajha_menu/sajha_menu/page/bar_order/bar_order.js
frappe.pages['bar-order'].on_page_load = function (wrapper) {
	new MyPage(wrapper);
}

MyPage = Class.extend({
	init: function (wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Bar Orders',
			single_column: true
		});

		frappe.realtime.on('kot_bot_refresh', ({ data }) => {
			this.refresh()
		});

		//make page
		this.filters();
		this.make();


	},

	refresh: async function () {

		// window.location.reload()
		let me = $(this);
		let company = this.page.fields_dict['company'].value
		let fromDate = this.page.fields_dict['fromDate'].value
		let toDate = this.page.fields_dict['toDate'].value
		let status = me[0].page.fields_dict['status'].value

		await frappe.call({
			method: "sajha_menu.sajha_menu.page.bar_order.bar_order.refreshKot",
			args: { "company": company, "fromDate": fromDate, "toDate": toDate, "status": status },
			callback: function (r) {
				$("#body-main")[0].innerHTML = frappe.render_template(r.message)
			}
		})
	},


	filters: async function () {
		let me = $(this);



		async function changesTab() {
			let company = me[0].page.fields_dict['company'].value
			let fromDate = me[0].page.fields_dict['fromDate'].value
			let toDate = me[0].page.fields_dict['toDate'].value
			let status = me[0].page.fields_dict['status'].value
			await frappe.call({
				method: "sajha_menu.sajha_menu.page.bar_order.bar_order.applyFilter",
				args: { "company": company, "fromDate": fromDate, "toDate": toDate, "status": status },
				callback: function (r) {
					$("#body-main")[0].innerHTML = frappe.render_template(r.message)
				}
			})


			// push dom element to page

		}
		await frappe.call({
			method: "sajha_menu.sajha_menu.page.bar_order.bar_order.getCompanyList",
			callback: function (r) {
				companyName = r.message
			}
		})

		this.page.add_field({
			label: 'Company',
			fieldtype: 'Link',
			fieldname: 'company',
			options: "Company",
			change() {
				changesTab();
			}
		});

		this.page.add_field({
			label: 'From Date',
			fieldtype: 'Date',
			fieldname: 'fromDate',
			default: frappe.datetime.get_today(),
			change() {
				changesTab();

			}
		});

		this.page.add_field({
			label: 'From Date',
			fieldtype: 'Date',
			fieldname: 'toDate',
			default: frappe.datetime.get_today(),
			change() {
				changesTab();
			}
		});

		this.page.add_field({
			label: 'Status',
			fieldname: 'status',
			fieldtype: "Select",
			options: [
				"Done",
				"In Process",
				"Pending"
			],
			change() {
				changesTab();
			}
		})


	},



	make: async function () {
		// grab the class 
		let me = $(this);

		let fromDate = frappe.datetime.get_today()
		let toDate = frappe.datetime.get_today()

		// body content
		let body = await frappe.call({
			method: "sajha_menu.sajha_menu.page.bar_order.bar_order.getBodyHTML",
			args: { "fromDate": fromDate, "toDate": toDate },
			callback: function (r) {
			}
		})
		// push dom element to page
		$(frappe.render_template(body.message, this)).appendTo(this.page.main)


		// execute methods
		// tableName();
	}

	// end of class
})